jerry-core/ecma/builtin-objects/ecma-builtin-json.cpp

```diff
@@ -0,0 +1,1815 @@
+        {
+          break;
+        }
+
+        /*
+         * We cannot optimize this function since any members
+         * can be changed (deleted) by the reviver function.
+         */
+
+        ecma_value_t value = ecma_get_completion_value_value (completion_value);
+
+        if (ecma_is_value_undefined (value))
+        {
+          ecma_completion_value_t comp_val = ecma_op_general_object_delete (object_p,
+                                                                            property_names_p[property_index],
+                                                                            false);
+          ecma_free_completion_value (comp_val);
+        }
+        else
+        {
+          ecma_builtin_json_define_value_property (object_p,
+                                                   property_names_p[property_index],
+                                                   value);
+        }
+
+        ecma_free_completion_value (completion_value);
+        completion_value = ecma_make_empty_completion_value ();
+      }
+
+      for (uint32_t i = 0; i < no_properties; i++)
+      {
+        ecma_deref_ecma_string (property_names_p[i]);
+      }
+      mem_heap_free_block (property_names_p);
+    }
+  }
+
+  /*
+   * Completition value can be an anything including exceptions.
+   */
+  if (ecma_is_completion_value_empty (completion_value))
+  {
+    ecma_value_t arguments_list[2];
+
+    arguments_list[0] = ecma_make_string_value (name_p);
+    arguments_list[1] = value;
+
+    completion_value = ecma_op_function_call (reviver_p,
+                                              ecma_make_object_value (holder_p),
+                                              arguments_list,
+                                              2);
+  }
+
+  ecma_free_value (value, true);
+  return completion_value;
+} /* ecma_builtin_json_walk */
+
+/**
+ * The JSON object's 'parse' routine
+ *
+ * See also:
+ *          ECMA-262 v5, 15.12.2
+ *
+ * @return completion value
+ *         Returned value must be freed with ecma_free_completion_value.
+ */
+static ecma_completion_value_t
+ecma_builtin_json_parse (ecma_value_t this_arg __attr_unused___, /**< 'this' argument */
+                         ecma_value_t arg1, /**< string argument */
+                         ecma_value_t arg2) /**< reviver argument */
+{
+  ecma_completion_value_t ret_value = ecma_make_empty_completion_value ();
+
+  ECMA_TRY_CATCH (string,
+                  ecma_op_to_string (arg1),
+                  ret_value);
+
+  ecma_string_t *string_p = ecma_get_string_from_value (string);
+  ecma_length_t length = (uint32_t) ecma_string_get_length (string_p);
+  size_t buffer_size = sizeof (lit_utf8_byte_t) * (length + 1);
+
+  lit_utf8_byte_t *str_start_p = (lit_utf8_byte_t*) mem_heap_alloc_block (buffer_size,
+                                                                  MEM_HEAP_ALLOC_SHORT_TERM);
+
+  if (str_start_p == NULL)
```

Comment:
The case is possible only if `buffer_size` is zero, not in case of out-of-memory.